src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimized.java

```diff
@@ -0,0 +1,330 @@
+
+            @Override
+            public void finishBundle(Context c) throws Exception {
+                if (null!=reader) {
+                    reader.close();
+                }
+                super.finishBundle(c);
+            }
+
+            private boolean accept(SAMRecord r, SimpleInterval region) {
+                if (r.getReadUnmappedFlag() ||
+                    r.getReferenceName() == null || r.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME) ||
+                    r.getAlignmentStart() == SAMRecord.NO_ALIGNMENT_START) {
+                    // unmapped read. We don't want it, *unless* their mate is in the region
+                    return (!r.getMateUnmappedFlag() && r.getMateAlignmentStart() >= region.getStart());
+                } else {
+                    // mapped read.
+                    // the query gives us reads that overlap the region. We want only reads that *start* in those intervals.
+                    return (r.getStart() >= region.getStart());
+                }
+            }
+        };
+    };
+
+    /**
+     * Given a shard that has reads and variants, query Google Genomics' Reference server and get reference info,
+     * and fill that and the correct variants into readContext.
+     */
+    public static DoFn<ContextShard, ContextShard> fillContext(final ReferenceDataflowSource refSource, final PCollectionView<SAMSequenceDictionary> refDictView, final GenomicsFactory.OfflineAuth auth) throws IOException {
+        return new DoFnWLog<ContextShard, ContextShard>("fillContext") {
+            private static final long serialVersionUID = 1L;
+            private transient Stopwatch getRef, queryVariants, sending, total;
+            private int elementCount;
+
+            @Override
+            public void startBundle(Context c) throws Exception {
+                super.startBundle(c);
+                getRef = Stopwatch.createUnstarted();
+                queryVariants = Stopwatch.createUnstarted();
+                sending = Stopwatch.createUnstarted();
+                total = Stopwatch.createUnstarted();
+                elementCount=0;
+            }
+
+            @Override
+            public void processElement(ProcessContext c) throws Exception {
+                total.start();
+                elementCount++;
+                ContextShard shard = c.element();
+                SAMSequenceDictionary refDict = c.sideInput(refDictView);
+                int margin = 500;
+                int start = shard.interval.getStart();
+                if (start>margin) start -= margin;
+                else start = 1; // it's 1-based, remember.
+                int end = shard.interval.getEnd();
+                getRef.start();
+                int len = refDict.getSequence(shard.interval.getContig()).getSequenceLength();
+                getRef.stop();
+                if (end<len-margin) end+= margin;
+                else end = len; // OK because those are 1-based closed intervals
```

Comment:
Well, the existing code protects us from overflows in case end is close to maxint. You're right, though, we can do without and min is easier to read.